### PR TITLE
Make FormBuilder useful to inherit

### LINKED ISSUE #3198: FormBuilder cannot be usefully inherited
`wagtail.wagtailforms.forms.FormBuilder` uses unbound references to refer to its methods, which means you can't inherit the form builder to override a method. Suggest using the bound methods.

It would also be good to abstract away common code like building the choices array, to allow this to be consistently overridden.

### PR DESCRIPTION
**Pull request by lb-**
Closes #3198

Changes the behaviour of `FormBuilder` to be more easily overridden and simpler to add fields without having to define new field types in multiple places.

Need to check we are happy with how this gets the methods that build the form fields, essentially assuming the methods will be written in a certain way.

* New tests added for this kind of behaviour, existing tests pass locally.
* Linting is a bit of a mess with the change to module names, getting hundreds of errors.
* Docs added for how to add custom fields.

### CODE REVIEW

File: wagtail/contrib/forms/forms.py [hunk 1]
```diff
@@ -77,33 +77,36 @@ def create_checkbox_field(self, field, options):
...
-        'dropdown': create_dropdown_field,
-        'multiselect': create_multiselect_field,
-        'radio': create_radio_field,
-        'checkboxes': create_checkboxes_field,
-        'checkbox': create_checkbox_field,
-        'hidden': create_hidden_field,
-    }
+    def get_create_field_function(self, type):
+        """
+            Takes string of field type returns a Django Form Field Instance.
```
Comment (tomdyson) [inline wagtail/contrib/forms/forms.py]: 'field type returns' -> 'field type and returns'

File: docs/reference/contrib/forms/customisation.rst [hunk 2]
```diff
@@ -604,3 +604,54 @@ Finally, we add a URL param of `id` based on the ``form_submission`` if it exist
...
             ], 'Email'),
         ]
+
+
+Adding Custom Field Types
+-------------------------
+
+Here is how you can add a custom field type, beyond the built in form field choices.
+
+Extend the default ``FORM_FIELD_CHOICES`` by importing and creating a new tuple that has our additional field added, the added items need to be tuples like ('field_type_key', 'Nice Field Label').
```
Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: 'field added, the' -> 'field added. The'

File: docs/reference/contrib/forms/customisation.rst [hunk 3]
```diff
@@ -604,3 +604,54 @@ Finally, we add a URL param of `id` based on the ``form_submission`` if it exist
...
+
+In your ``FormField`` model definition, override the ``field_type`` so that we can add the additional field choice to the dropdown.
+
+Create a new class that extends ``FormBuilder``, this will have a method to match the new field type we want to make available, this method must return an instanced Django Form Field.
```
Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: 'available, this method must' -> 'available, and must'

Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: 'an instanced Django Form Field' - is this clear to everyone? It wasn't to me.

Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: 'Django Form Field' -> 'Django form field'

**Comment by lb-** [comment]
@tomdyson thanks for the feedback, I will get some updates through.

**Comment by lb-** [comment]
@tomdyson let me know if these docs make more sense.
@danni would love to get your thoughts also on the docs.

File: docs/reference/contrib/forms/customisation.rst [hunk 4]
```diff
@@ -606,48 +606,58 @@ Finally, we add a URL param of `id` based on the ``form_submission`` if it exist
...
-In your ``FormField`` model definition, override the ``field_type`` so that we can add the additional field choice to the dropdown.
 
-Create a new class that extends ``FormBuilder``, this will have a method to match the new field type we want to make available, this method must return an instanced Django Form Field.
-The name of this method must match the field_type_key for the additional choice added, the format is ``create_<field_type_key>_field``.
+Then, create and use a new form builder class.
 
-Add the ``form_builder`` attribute to your ``FormPage`` method so that this extended form builder is used.
+* Define a new form builder class that extends the ``FormBuilder`` class.
+* Add a method that will return a created Django form field for the new field type.
+* Its name must be in the format: ``create_<field_type_key>_field``. eg. ``create_slug_field``
```
Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: '. eg. ' -> ', e.g. '

File: docs/reference/contrib/forms/customisation.rst [hunk 5]
```diff
@@ -606,48 +606,58 @@ Finally, we add a URL param of `id` based on the ``form_submission`` if it exist
...
-Adding Custom Field Types
--------------------------
+Adding a custom field type
+--------------------------
 
-Here is how you can add a custom field type, beyond the built in form field choices.
+First, make the new field type available in the page editor by changing your ``FormField`` model.
 
-Extend the default ``FORM_FIELD_CHOICES`` by importing and creating a new tuple that has our additional field added, the added items need to be tuples like ('field_type_key', 'Nice Field Label').
+* Create a new set of choices which includes the original ``FORM_FIELD_CHOICES`` along with new field types you want to make avaialable.
```
Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: avaialable -> available

File: docs/reference/contrib/forms/customisation.rst [hunk 6]
```diff
@@ -606,48 +606,58 @@ Finally, we add a URL param of `id` based on the ``form_submission`` if it exist
...
+* Each choice must contain a unique key and a human readable name of the field. eg. ``('slug', 'URL Slug')``
```
Comment (tomdyson) [inline docs/reference/contrib/forms/customisation.rst]: '. eg.' -> ', e.g.'

File: wagtail/contrib/forms/tests/test_forms.py [hunk 7]
```diff
@@ -162,3 +163,51 @@ def test_fields(self):
...
+    def test_using_custom_form_builder(self):
+        """Tests that charfield max_length is 120 characters."""
+        form_class = self.form_page.get_form_class()
+        form = form_class()
+        # check name field exists
+        self.assertIsInstance(form.base_fields['name'], forms.CharField)
+        # check max_length is set
+        self.assertEqual(form.base_fields['name'].max_length, 120)
+
+
```
Comment (kaedroho) [inline wagtail/contrib/forms/tests/test_forms.py]: Extra newline here (flake8 should've picked this up)

File: wagtail/contrib/forms/forms.py [hunk 8]
```diff
@@ -77,33 +77,36 @@ def create_checkbox_field(self, field, options):
...
+        if create_field_function:
+            return create_field_function
+        else:
+            import inspect
+            method_list = [
+                f[0] for f in
+                inspect.getmembers(self.__class__, inspect.isfunction)
+                if f[0].startswith('create_') and f[0].endswith('_field')
+            ]
+            raise Exception(
```
Comment (kaedroho) [inline wagtail/contrib/forms/forms.py]: Should this be an ``AttributeError`` instead?

**Comment by lb-** [comment]
@kaedroho thanks for picking those up, have updated the PR

**Comment by lb-** [comment]
@tomdyson let me know if you have any issues/other feedback, otherwise I will look at merging this in later this week.

**Comment by lb-** [comment]
Merged in https://github.com/wagtail/wagtail/commit/3eac584c7c6a73db9cf8c238c806357d1f3b2c3e

**Pull request closed without merging**